processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java

```diff
@@ -49,16 +51,41 @@
  */
 public class MapperConfiguration {
 
+    private static Map<Element, MapperConfiguration> configurations = new WeakHashMap<Element, MapperConfiguration>();
+
     private final MapperPrism mapperPrism;
+    private final AnnotationMirror componentModelAnnotation;
     private final MapperConfigPrism mapperConfigPrism;
     private final DeclaredType config;
 
-    public static MapperConfiguration getInstanceOn(Element e) {
-        return new MapperConfiguration( MapperPrism.getInstanceOn( e ) );
+    private final FormattingMessager messager;
+
+    public static MapperConfiguration getInstanceOn(Element e, FormattingMessager messager) {
+        // store configurations per Element instance in a cache map
+        // this prevents creating many instances of this configuration for the same Element
+        // it will be assumed that this method will never be called with different messagers for the same Element
+        // one reason to cache instances is that the the MapperConfiguration validates the Element and multiple
+        // instantiations would lead to multiple identical warnings/errors
+
+        MapperConfiguration config = configurations.get( e );
+        if ( config == null ) {
+            config = new MapperConfiguration( e, messager );
+            configurations.put( e, config );
+        }
+
+        return config;
+    }
+
+    private static AnnotationMirror findComponentModelAnnotation(Element e, FormattingMessager messager) {
+        AnnotationMirror componentModelAnnotation = null;
+
+
+        return componentModelAnnotation;
     }
 
-    private MapperConfiguration(MapperPrism mapperPrism) {
-        this.mapperPrism = mapperPrism;
+    private MapperConfiguration(Element e, FormattingMessager messager) {
+        this.mapperPrism = MapperPrism.getInstanceOn( e );
```

Comment:
Moved the `MapperPrism.getInstanceOn` to the constructor as the `Element` is required a few lines later. I also added the `FormattingMessager` to be able to print the validation messages.